cpp/src/arrow/engine/expression.h

```diff
@@ -0,0 +1,490 @@
+    TABLE,
+  };
+
+  /// Construct a Scalar type.
+  static ExprType Scalar(std::shared_ptr<DataType> type);
+  /// Construct an Array type.
+  static ExprType Array(std::shared_ptr<DataType> type);
+  /// Construct a Table type.
+  static ExprType Table(std::shared_ptr<Schema> schema);
+
+  /// \brief Shape of the expression.
+  Shape shape() const { return shape_; }
+
+  /// \brief DataType of the expression if a scalar or an array.
+  /// WARNING: You must ensure the proper shape before calling this accessor.
+  const std::shared_ptr<DataType>& data_type() const { return data_type_; }
+  /// \brief Schema of the expression if of table shape.
+  /// WARNING: You must ensure the proper shape before calling this accessor.
+  const std::shared_ptr<Schema>& schema() const { return schema_; }
+
+  /// \brief Indicate if the type is a Scalar.
+  bool IsScalar() const { return shape_ == SCALAR; }
+  /// \brief Indicate if the type is an Array.
+  bool IsArray() const { return shape_ == ARRAY; }
+  /// \brief Indicate if the type is a Table.
+  bool IsTable() const { return shape_ == TABLE; }
+
+  template <Type::type TYPE_ID>
+  bool IsTypedLike() const {
+    return (IsScalar() || IsArray()) && data_type_->id() == TYPE_ID;
+  }
+
+  /// \brief Indicate if the type is a predicate, i.e. a boolean scalar.
+  bool IsPredicate() const { return IsTypedLike<Type::BOOL>(); }
+
+  /// \brief Cast to DataType/Schema if the shape allows it.
+  Result<ExprType> CastTo(const std::shared_ptr<DataType>& data_type) const;
+  Result<ExprType> CastTo(const std::shared_ptr<Schema>& schema) const;
+
+  /// \brief Broadcasting align two types to the largest shape.
+  ///
+  /// \param[in] lhs, first type to broadcast
+  /// \param[in] rhs, second type to broadcast
+  /// \return broadcasted type or an error why it can't be broadcasted.
+  ///
+  /// Broadcasting promotes the shape of the smallest type to the bigger one if
+  /// they share the same DataType. In functional pattern matching it would look
+  /// like:
+  ///
+  /// ```
+  /// Broadcast(rhs, lhs) = match(lhs, rhs) {
+  ///   case: ScalarType(t1), ScalarType(t2) if t1 == t2 => ScalarType(t)
+  ///   case: ScalarType(t1), ArrayType(t2)  if t1 == t2 => ArrayType(t)
+  ///   case: ArrayType(t1),  ScalarType(t2) if t1 == t2 => ArrayType(t)
+  ///   case: ArrayType(t1),  ArrayType(t2)  if t1 == t2 => ArrayType(t)
+  ///   case: _ => Error("Types not compatible for broadcasting")
+  /// }
+  /// ```
+  static Result<ExprType> Broadcast(const ExprType& lhs, const ExprType& rhs);
+
+  bool Equals(const ExprType& type) const;
+
+  std::string ToString() const;
+
+  ExprType(const ExprType& copy);
+  ExprType(ExprType&& copy);
+  ~ExprType();
+
+ private:
+  /// Table constructor
+  ExprType(std::shared_ptr<Schema> schema, Shape shape);
+  /// Scalar or Array constructor
+  ExprType(std::shared_ptr<DataType> type, Shape shape);
+
+  union {
+    // Zero initialize the pointer or Copy/Assign constructors will fail.
```

Comment:
It's a union quirk.

Reply:
`util::variant` _is_ arcane. I initially used variant until I switched the getter for type/schema to return a const ref instead of a copy. You proposed to use static shared_ptr when the type is invalid so we can still return a const-ref.

I think I'll go with util::variant just for the sake of consistency in the code base.